apps/ejabberd/src/ejabberd_sm.erl

```diff
@@ -202,7 +215,8 @@ get_user_resources(User, Server) ->
 
 -spec get_session_ip(User :: ejabberd:user(),
```

Comment:
I would opt for the following syntax in situations where the spec/return type is so long that it has to be wrapped on arrow:

```erlang
-spec get_session_ip(User, Server, Resource) -> Result when
      User :: ejabberd:user(),
      Server :: ejabberd:server(),
      Resource :: ejabberd:resource()
      Result :: undefined | {inet:ip_address(), integer()}.
```

This way we have plenty of space for each parameter type and readability of the `-> some_return_type()` is not degraded.
For multiple alternatives which don't make it into a single line we can then break on `|` like this:

```erlang
      Result :: undefined
              | {inet:ip_address(), integer()}.
```